### ISSUE TITLE
Spotify isn't playing anything

### ISSUE THREAD

**Issue by helenaballentine** [futo-org/grayjay-android#1588]
### What happened?

Whenever I try to play a podcast on Spotify, it loads for a brief second before it just gives up. There is no error message, it just stops loading and a play button shows that you can't do anything with. I tried logging out of Spotify, but it says you need to be logged in to play anything. I've also tried clearing the cache and cookies, but that didn't change anything. I am on Android 9 and my phone won't let me upgrade past that.

### Grayjay Version

268

### What plugins are you seeing the problem on?

Other

### Plugin Version

12

### When do you experience the issue?

- [X] While logged in
- [ ] While logged out
- [ ] N/A

**Comment by b-risk** [comment]
This is an odd bug I've experienced as well since the plugin was introduced, the temporary fix that has worked for me is just re-logging in, but I suspect there is something that could be added to check if this bug happens though I'm not sure the cause.

**Comment by helenaballentine** [comment]
> This is an odd bug I've experienced as well since the plugin was introduced, the temporary fix that has worked for me is just re-logging in, but I suspect there is something that could be added to check if this bug happens though I'm not sure the cause.

I tried logging out and back in, but it unfortunately didn't change anything.

**Comment by kaidelorenzo** [comment]
@helenaballentine does playing songs work? has playing podcasts ever worked? would you be comfortable sharing an episode that isn't working for you?

**Comment by helenaballentine** [comment]
> @helenaballentine does playing songs work? has playing podcasts ever worked? would you be comfortable sharing an episode that isn't working for you?

I just tested it and songs do not work. Podcasts were working for quite a while and then it stopped I think about a month ago. An example of one not working is the channel Rotten Mango and the episode is "405 Man From Viral TikTok Looks Exactly Like Missing Husband - Did He Fake His Own Death For Another Girl"

**Comment by codebymatter** [comment]
I have been using Spotify plugin since it's release andbsubsequent version 2 update. Other thqn not being able to Import playlists or subscriptions and the occasional music or podcast track not working, my experence with Spotify has been very good.

Regarding songs/albums not playing, I can easily provide an example album to see for yourself that it does not play. Lookup the early-2000s album, "Audioslave" and try playing any track. Will not play. I can listen to the full album on YT so no huge loss but thought I would give input.

**Comment by b-risk** [comment]
> > This is an odd bug I've experienced as well since the plugin was introduced, the temporary fix that has worked for me is just re-logging in, but I suspect there is something that could be added to check if this bug happens though I'm not sure the cause.
> 
> I tried logging out and back in, but it unfortunately didn't change anything.

Can you log into Spotify on a browser and play songs from there?

**Comment by helenaballentine** [comment]
> > > This is an odd bug I've experienced as well since the plugin was introduced, the temporary fix that has worked for me is just re-logging in, but I suspect there is something that could be added to check if this bug happens though I'm not sure the cause.
> > 
> > 
> > I tried logging out and back in, but it unfortunately didn't change anything.
> 
> Can you log into Spotify on a browser and play songs from there?

Yes I can. I played the song Brain Rot by the artist Nihmune and I also played the same podcast episode I mentioned above on the Spotify website. Both played with no issues.

**Comment by SovietGP** [comment]
> > > > This is an odd bug I've experienced as well since the plugin was introduced, the temporary fix that has worked for me is just re-logging in, but I suspect there is something that could be added to check if this bug happens though I'm not sure the cause.
> > > 
> > > 
> > > I tried logging out and back in, but it unfortunately didn't change anything.
> > 
> > 
> > Can you log into Spotify on a browser and play songs from there?
> 
> Yes I can. I played the song Brain Rot by the artist Nihmune and I also played the same podcast episode I mentioned above on the Spotify website. Both played with no issues.

I get the same issue, Spotify App plays music, 1DM, Firefox and Vivaldi all play music except for Grayjay. Grayjay shows Spotify Playlists and songs but it won't play anything, it starts to buffer then you see a play icon which does nothing.

**Comment by kaidelorenzo** [comment]
I've made partial progress.

The reason why some songs don't play is because of some odd file structure at Spotify these changes will fix that issue:
https://gitlab.futo.org/videostreaming/plugins/spotify/-/merge_requests/2
https://gitlab.futo.org/videostreaming/grayjay/-/merge_requests/68

They should release with the next version of Grayjay.

The issue where some devices can't play anything is still a mystery. It might have to do with some device specific Widevine issue. I added code in the Grayjay changes linked above that checks to see if the device supports Widevine. Potentially that will shed some light.

**Comment by SovietGP** [comment]
Updated to the latest version, same issue occurs. I didn't see anything on if my device supports Widevine.

**Comment by kaidelorenzo** [comment]
Yeah this most recent update was limited in scope and didn't include the Spotify changes

**Comment by SovietGP** [comment]
My widevine vendor is Google ver 15, L3

**Comment by SovietGP** [comment]
has anyone got an6 updates on why it's not working?

**Comment by kaidelorenzo** [comment]
@SovietGP what's your phone and OS? Do you have the bootloader unlocked or the OS rooted? I've figured out the issue with some songs not playing but it's still a mystery for the issue of nothing playing.

**Comment by helenaballentine** [comment]
> @SovietGP what's your phone and OS? Do you have the bootloader unlocked or the OS rooted? I've figured out the issue with some songs not playing but it's still a mystery for the issue of nothing playing.

I have a Sony XZ2 and Android 9. My phone is unlocked and not rooted.

**Comment by kaidelorenzo** [comment]
the bootloader is unlocked? do you have a custom rom installed?

**Comment by helenaballentine** [comment]
Took me a minute to find out, but my bootloader is not unlocked. I don't have any custom ROM installed.

**Comment by kaidelorenzo** [comment]
Maybe it's an android 9 issue. I'll look into that.

**Comment by anuragtyagi-dev** [comment]
> Maybe it's an android 9 issue. I'll look into that.

Yes. Same Issue I've encountered and flagged today. I've been using Grayjay over 2 months and the previous version of Spotify Plugins were working fine for me but after the version maybe 11 or this current 12 , nothing seems to be playing. It just shows PAUSE/PLAY button.
Android : 9
Phone : RealmeXT RMX1921

**Comment by kaidelorenzo** [comment]
Can you try playing from the browser on your phone ideally Google Chrome and let me know if that works?

https://open.spotify.com/album/600XgAY1N6sRPVczmBogiF

**Comment by helenaballentine** [comment]
Yes, it does. I've tested it before and both podcasts and music play in the Spotify website. Just to be sure, I clicked the link you sent and it did play with no issues in Chrome.

**Comment by kaidelorenzo** [comment]
Just to clear Chrome on the same android phone?

**Comment by helenaballentine** [comment]
Yes. The Spotify website on Chrome using the same phone.

**Comment by kaidelorenzo** [comment]
@SovietGP could you try playing in the browser on your android 10 device?

**Comment by anuragtyagi-dev** [comment]
> @kaidelorenzo Plays fine after I allow DRM content, using firefox nightly.

What you mean ?

**Comment by anuragtyagi-dev** [comment]
@SovietGP Yes the dev's have actually acknowledged the issue with YouTube Plugin as well that has " Block Detected " error and videos are not playing. Some way around this is I found , that switching between mobile internet and Wifi solves this somehow. And for the Spotify issue , Yes Spotify Official Apk ( Playstore ) is working like it should and even Spotify on Web ( Tried on Chrome , Brave , FireFox ) is working fine as well , So the issue with Spotify Plugin on " Android 9 "  is of Plugin itself.

**Comment by SovietGP** [comment]
I'm on A10 and have this issue, YouTube playback now works after update and age restricted contenr load fine. Spotify still won't play anything I haven't noticed any changes.

**Comment by anuragtyagi-dev** [comment]
Some Youtube videos still not loading and player showing and stuck on Play/Pause button but otherwise mostly works fine now , also spotify still not working 
Android 9, RealmeXT

**Comment by helenaballentine** [comment]
Just wanted to update. I'm now on Grayjay 

> /version 281 and Spotify still doesn't work. I am getting this error when I try to play something on it now though.

![Image](https://github.com/user-attachments/assets/41c9c629-3922-4f59-9bbb-183b60e936c2)

**Comment by pixal6191** [comment]
Having the same issue here as well, the song is just stuck at the pause screen

I'm using the Play Store version so dunno if it's something to do with that.

Android version: 11
App version: 281
Patch version: 12
Device: realme c3

**Comment by kaidelorenzo** [comment]
@pixal6191 you can't play any songs? Can you play Spotify in the browser on that phone?

**Comment by kaidelorenzo** [comment]
Can you check a song from this album?

https://open.spotify.com/album/600XgAY1N6sRPVczmBogiF

**Comment by pixal6191** [comment]
Kind of stupid of me but i wasn't logged in the entire time, but when i did, this popped up ![Screenshot_2025-02-08-22-26-02-98_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://github.com/user-attachments/assets/f9009a10-9efe-4282-a578-155242e8b220)

**Comment by kaidelorenzo** [comment]
like when you tried to log in to spotify on grayjay? that's kind of what i was expecting you to get when you opened spotify in a mobile browser

**Comment by pixal6191** [comment]
This was in the browser

**Comment by kaidelorenzo** [comment]
So you're not able to play spotify in the browser on that phone?

**Comment by pixal6191** [comment]
Yeah and it shows that when I'm logged in

**Comment by kaidelorenzo** [comment]
Just FYI we're not aiming to support Spotify in Grayjay for devices that can't play in the browser.

**Comment by anuragtyagi-dev** [comment]
What kind of error is this ? music is playing while < not logged in > but getting this " block protected or private browsing error " just after logging-In with account. What's the reason and how it's " mobile that can't play spotify in browser " related ?

**Comment by kaidelorenzo** [comment]
I'm not sure exactly what you're saying. My understanding is that Spotify doesn't allow you to play anything without logging in.

Additionally from my testing failure to play most often comes down to widevine issues.

Could you rephrase what your question is?

**Comment by anuragtyagi-dev** [comment]
@kaidelorenzo 
> Could you rephrase what your question is?

Sorry , I was asking in reference with your quote reply , You said " we're not aiming to support Spotify in Grayjay for devices that can't play in the browser."  When I tried to play song or any playlist in spotify on chrome or other browser without logging in , same error pops up that @pixal6191 showed with their screenshot, but when I logged in with my account any song or playlist is playing fine. 
My question was , how this is related with " Mobile that can't play spotify in browser " , Is this our Mobile's Android Version related or something else.

My Device :
Realme XT Version : RMX1921EX_11_A.20 
Android: 9
( Spotify Plugin in Grayjay is not working with logged-In and logged-Out , but Spotify is working fine in browser while logged-In ( chrome , brave ) but not working while logged-out , also Spotify app is working fine as well. )

**Comment by kaidelorenzo** [comment]
oh interesting. so it only fails to play when playing preview tracks. if it plays while logged in we should be able to support that. the reason why i said that is that we use the same playing approach and the web app. so if the web app doesn't work in browser that it definitely won't work in Grayjay.

**Comment by anuragtyagi-dev** [comment]
I get it now, it's obvious for Grayjay to support Spotify only if Spotify works like it should in web app. and yes the web app is working fine while only logged-In , the playlist you mentioned is playing fine ( tried on chrome and brave browser ). I don't know what's the issue while logged-out.

**Comment by SovietGP** [comment]
Spotify works fine on all my browsers except for Grayjay. Grayjay shows a DRM error when I view the page, without playing anything. 

![Image](https://github.com/user-attachments/assets/0bfda697-8600-4b7d-8458-4e8c82488ba8)

**Comment by SovietGP** [comment]
> Spotify works fine on all my browsers except for Grayjay. Grayjay shows a DRM error when I view the page, without playing anything.
> 
> ![Image](https://github.com/user-attachments/assets/0bfda697-8600-4b7d-8458-4e8c82488ba8)

I get a Widevine error if I use a VPN and relog into Spotify in the Grayjay app. 

![Image](https://github.com/user-attachments/assets/cff92a8d-492d-441a-80e9-2596838bc54c)

**Comment by kaidelorenzo** [comment]
> Spotify works fine on all my browsers except for Grayjay. Grayjay shows a DRM error when I view the page, without playing anything. 
> 
> ![Image](https://github.com/user-attachments/assets/0bfda697-8600-4b7d-8458-4e8c82488ba8)

Is this when logging into Spotify in Grayjay? That's not Grayjay UI

**Comment by SovietGP** [comment]
@kaidelorenzo It is infact the Grayjay UI, I am logged into Spotify that works fine, it's when I go to a song and tap settings and open the "Page" option, that is when it loads the internal browser that Grayjay uses and gives me this DRM error. 

![Image](https://github.com/user-attachments/assets/4e4054cb-bab9-4876-bb39-434598472faa)

**Comment by SovietGP** [comment]
> > Spotify works fine on all my browsers except for Grayjay. Grayjay shows a DRM error when I view the page, without playing anything.
> > ![Image](https://github.com/user-attachments/assets/0bfda697-8600-4b7d-8458-4e8c82488ba8)
> 
> I get a Widevine error if I use a VPN and relog into Spotify in the Grayjay app.
> 
> ![Image](https://github.com/user-attachments/assets/cff92a8d-492d-441a-80e9-2596838bc54c)

I fixed this error, Spotify reset my pass for many Logins.

**Comment by SovietGP** [comment]
It seems to me that Grayjay disables DRM in its browser and that causes no audio playback and the blocked content error.

**Comment by anuragtyagi-dev** [comment]
@SovietGP Don't Panic , They also have life.

**Comment by SovietGP** [comment]
@anuragtyagi-dev I'm not panicking, I am curious to know if any progress has being made, I completely understand people need time for themselves and have a personal life to attend to. It is not my insion to push people, some time pasted since there was any updates, that is all.

**Comment by SovietGP** [comment]
Wae any progress made with this issue?

**Comment by kaidelorenzo** [comment]
No progress on android 9/10 issues (i need to get around to finding a device i can replicate the issue on). as stated above the issue with not all songs being playable has a fix that is still not released.

**Comment by kaidelorenzo** [comment]
hmm interesting. is chrome fully up to date?

**Comment by kaidelorenzo** [comment]
yeah so that issue then is that widevine has the option to require host verification. and that i think uses play integrity. spotify seems to use that option. currently we don't support phones that can't play spotify in browser.

**Comment by SovietGP** [comment]
I get the same issue like @pauel3312, but I'm on a different OS; MIUI 12 eu, Firefox and Chrome works but Vivaldi does not. It may be an integrity issue but I'm not 100% sure.

**Comment by kaidelorenzo** [comment]
You can play past like 10 seconds in chrome and Firefox? And that's android 11?

**Comment by SovietGP** [comment]
@kaidelorenzo No I cannot it stops indeed at 10sec. So in your opinion is there a fix or a workaround for any OS and device with this issue? Or do we have to deal with it until maybe someone finds a way or something changes?

**Comment by kaidelorenzo** [comment]
The only solution would be to implement playing another kind of Spotify media type (non widevine). We're considering doing that but haven't made any decisions on it

**Comment by SovietGP** [comment]
@kaidelorenzo It could be risky depending where you are and your laws, so I recommend you check your laws on this, incase grayjay gets flagged and removed. Thank you for your help, I can manage without Spotify in Grayjay for the time being.

**Comment by helenaballentine** [comment]
Just wondering if there's any update on this. No worries if there isn't.

**Comment by dominmana** [comment]
i have lineageos  21 with microg on two devices:

- sony xperia xa2 (pioneer)
- motorola moto g 5g plus (nairo)

i have grayjay running on both but on the xperia the spotify plugin has the broken play button (as described in this thread) while on the motorola it totally works.
same versions, same setup, different behaviors

i hope this can help narrow it down

**Comment by dominmana** [comment]
> i have lineageos 21 with microg on two devices:
> 
> * sony xperia xa2 (pioneer)
> * motorola moto g 5g plus (nairo)
> 
> i have grayjay running on both but on the xperia the spotify plugin has the broken play button (as described in this thread) while on the motorola it totally works. same versions, same setup, different behaviors
> 
> i hope this can help narrow it down

let me just add what i mean by this:
it can't be a lineageos issue because i have the same system on both phones ani the plugin works on motorola, maybe it has something to do with the phone architecture or drivers? someone else mentioned another sony device with this issue.

**Comment by admintm** [comment]
Had this issue since i tried spotify add-on in grayjay on my Sony Xperia 1 V with lineageOS 21 for the first time almost a year ago. It never worked no matter what i tried: sign in/out, clear cookies, reinstall, update grayjay, update the plugin, connecting through "dirty" cellular ip or clean dedicated static public ip etc. etc. 
Every now and then when i tried playing content from spotify even the entire grayjay app crashed. The crashdump always mentioned something about DRM, so it is likely a widevine issue. (I always sent the dump to the developers.)

On the same device in the waterfox browser spotify plays fine after allowing the site to "play DRM protected content". Also no 10s limmit on the songs, everything works just fine. I assume this will be the same on other firefox flavours. 
But any browser (or any app i guess) that uses the builtin android webview is a diffrent storry, for example trying spotify in the browser included with lineageOS (a basic webview wrapper) does not work at all. It comes up with "Playback not supported because of DRM blablabla..."

Back then i decided to stop trying and used spotube instead as a spotify client, because it just worked. It streames the media from youtube so no need for widevine DRM bullshit. Now that spotube has been shut down (cease and desisted by spotify - @FUTO please send your lawyers to defend that dev. and maybe take over the project) i'm back in grayjay trying to get my beloved and flawlessly working youtube client to also become my spotify client with no success so far. 

Further testing also showed that the grayjay spotify add-on works just fine on another android device with stock android. So likely the system webview needs at least widevine L3 support and likely lineageOS does not bake that in to their webview. Browsers that implement their own version of widevine L3 on top of what the system offers also work fine. Anything beyond L3 widevine (L2 or L1) will probably never work on a device with unloked bootloader, i do understand that. But since it seems like spotify is happy with L3 widevine couldn't grayjay "just" implement that within the app like "***insert-your-favourite-flavour-of***" fox can?

**Comment by Liliashepherd** [comment]
Updated to the latest version, same issue occurs. I didn't see anything on if my device supports Widevine.

**Comment by Zvonimir-FUTO** [comment]
We will make a proper announcement later but TLDR for now, due to Spotify cracking down on 3rd party clients we will drop maintenance of the Spotify plugin for foreseeable future.

For the devs, feel free to fork the plugin and patch it yourself.
